sections/Chapter 5 Sprint 3/db_design_redone.tex

```diff
@@ -0,0 +1,148 @@
+
+        foreach (WordData articleWord in article.Words)
+        {
+            WordListModel wordListModel = JsonDbUtility
+                .WordDataToWordList(articleWord);
+            AppearsInModel appearsInModel = JsonDbUtility
+                .ArticleWordDataToAppearsIn(article, articleWord);
+
+            if (dbContext.Wordlist.Find(wordListModel.WordName) == null)
+            {
+                words.Add(wordListModel);
+            }
+
+            appearsInModels.Add(appearsInModel);
+        }
+
+        dbContext.Wordlist.AddRange(words);
+        dbContext.FileList.Add(fileListModel);
+    }
+        
+    dbContext.SaveChanges();
+    dbContext.AppearsIn.AddRange(appearsInModels);
+    dbContext.SaveChanges();
+
+    Console.WriteLine($"Added {articles.Length} entries.");
+
+    return Ok(message == string.Empty ? "Ok" : message);
+}
+\end{lstlisting}
+
+It proved difficult to work with and caused more problems than benefits, especially under the aforementioned time constraints. 
+Therefore, we decided to rework the design completely by creating C\# models that would translate to a relational database design using EF Core migrations.
+The POST method in the WordCount controller for the redesigned database can be seen in code snippet \ref{lst:new_controller}.
+
+
+\begin{lstlisting}[language=CSharp, caption={New controller}, label={lst:new_controller}]
+[HttpPost]
+public IActionResult Post([FromBody] JsonElement jsonElement)
+{
+    JsonSchemaModel? schema = unitOfWork.SchemaRepository
+        .Find(s => s.SchemaName == WordCountSchemaName);
+    string jsonInput = jsonElement.GetRawText();
+
+    if (schema == null)
+    {
+        return StatusCode(500, $"\"{WordCountSchemaName}\" schema does not exist.");
+    }
+
+    // Get schema and use for validating
+    if (!new JsonValidator<ArticleJsonModel[]>(schema.JsonString)
+            .IsValid(jsonInput, 
+                out ArticleJsonModel[] jsonArticles))
+    {
+        return BadRequest("Wrong body syntax, does not follow schema.");
+    }
+
+    IEnumerable<Article> result = RemoveDuplicates(jsonArticles, out StringBuilder message);
+
+    // Insert article
+    IEnumerable<Article> enumerable = result as Article[] 
+        ?? result.ToArray();
+
+    unitOfWork.ArticleRepository.Insert(enumerable);
+
+    foreach (Article article in enumerable)
+    {
+        Console.WriteLine($"Added {article.Title}");
+    }
+
+    return Ok(message.ToString());
+}
+\end{lstlisting}
+    
+By comparing the semantics and overall size of code snippet \ref{lst:old_controller} and \ref{lst:new_controller}, it should be apparent that the redesigned database is much easier to work with using EF Core, as it removed a lot of unnecessary complexity in the code.
+However, despite the many advantages from the redesign, it has a major memory issue.
+In our rush to quickly implement the redesign, we unfortunately overlooked how to effectively store words from articles. 
```

Comment:
```suggestion
In our rush to quickly implement the redesign, we, unfortunately, overlooked how to effectively store words from articles. 
```

Reply:
Not an interposed sentence so no commas.